File: app/api/vault.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
)

type FileInfo struct {
	Name       string `json:"name"`
	Path       string `json:"path"`
	Size       int64  `json:"size"`
	Modified   int64  `json:"modified"`
	Type       string `json:"type"`
	IsMarkdown bool   `json:"is_markdown"`
	IsCSV      bool   `json:"is_csv"`
}

type FileResult struct {
	Path    string `json:"path,omitempty"`
	Name    string `json:"name,omitempty"`
	Content string `json:"content,omitempty"`
	Size    int64  `json:"size,omitempty"`
	Error   string `json:"error,omitempty"`
}

type WriteResult struct {
	Success      bool   `json:"success"`
	Path         string `json:"path,omitempty"`
	BytesWritten int    `json:"bytes_written,omitempty"`
	Mode         string `json:"mode,omitempty"`
	Error        string `json:"error,omitempty"`
}

type SearchResult struct {
	Path      string `json:"path"`
	MatchType string `json:"match_type"`
	Line      int    `json:"line,omitempty"`
	Snippet   string `json:"snippet"`
}

type VaultManager struct {
	RootDir string
}

func NewVaultManager(rootDir string) *VaultManager {
	abs, err := filepath.Abs(rootDir)
	if err != nil {
		abs = rootDir
	}
	return &VaultManager{RootDir: abs}
}

func (v *VaultManager) resolveSafePath(relPath string) (string, error) {
	clean := strings.TrimLeft(filepath.Clean(relPath), "/\\.")
	target := filepath.Join(v.RootDir, clean)
	absTarget, err := filepath.Abs(target)
	if err != nil {
		return "", err
	}

	if !strings.HasPrefix(absTarget, v.RootDir) {
		return "", errors.New("access denied: path is outside vault root")
	}
	return absTarget, nil
}

func (v *VaultManager) ListFiles(subpath string, maxDepth int) ([]FileInfo, error) {
	targetDir, err := v.resolveSafePath(subpath)
	if err != nil {
		return nil, err
	}

	var results []FileInfo
	baseParts := len(strings.Split(filepath.ToSlash(targetDir), "/"))

	err = filepath.Walk(targetDir, func(p string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}

		parts := len(strings.Split(filepath.ToSlash(p), "/"))
		if parts-baseParts > maxDepth {
			if info.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}

		name := info.Name()
		if info.IsDir() {
			if strings.HasPrefix(name, ".") || name == "__pycache__" || name == "node_modules" {
				return filepath.SkipDir
			}
			return nil
		}

		if strings.HasPrefix(name, ".") {
			return nil
		}

		rel, err := filepath.Rel(v.RootDir, p)
		if err != nil {
			return nil
		}

		ext := strings.ToLower(filepath.Ext(p))
		results = append(results, FileInfo{
			Name:       name,
			Path:       filepath.ToSlash(rel),
			Size:       info.Size(),
			Modified:   info.ModTime().Unix(),
			Type:       strings.TrimPrefix(ext, "."),
			IsMarkdown: ext == ".md",
			IsCSV:      ext == ".csv",
		})
		return nil
	})

	return results, err
}

func (v *VaultManager) ReadFile(relPath string, maxBytes int) FileResult {
	target, err := v.resolveSafePath(relPath)
	if err != nil {
		return FileResult{Error: err.Error()}
	}

	info, err := os.Stat(target)
	if err != nil {
		return FileResult{Error: fmt.Sprintf("File '%s' does not exist.", relPath)}
	}
	if info.IsDir() {
		return FileResult{Error: fmt.Sprintf("'%s' is a directory, not a file.", relPath)}
	}

	file, err := os.Open(target)
	if err != nil {
		return FileResult{Error: err.Error()}
	}
	defer file.Close()

	if maxBytes <= 0 {
		maxBytes = 100000
	}

	buf := make([]byte, maxBytes)
	n, err := io.ReadFull(file, buf)
	if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
		return FileResult{Error: err.Error()}
	}

	content := string(buf[:n])
	if int64(n) < info.Size() {
		content += fmt.Sprintf("\n\n... [Content truncated, total %d bytes] ...", info.Size())
	}

	return FileResult{
		Path:    filepath.ToSlash(relPath),
		Name:    info.Name(),
		Content: content,
		Size:    info.Size(),
	}
}

func (v *VaultManager) WriteFile(relPath string, content string, appendMode bool) WriteResult {
	target, err := v.resolveSafePath(relPath)
	if err != nil {
		return WriteResult{Success: false, Error: err.Error()}
	}

	if err := os.MkdirAll(filepath.Dir(target), 0755); err != nil {
		return WriteResult{Success: false, Error: err.Error()}
	}

	flag := os.O_CREATE | os.O_WRONLY
	modeStr := "overwrite"
	if appendMode {
		flag |= os.O_APPEND
		modeStr = "append"
	} else {
		flag |= os.O_TRUNC
	}

	f, err := os.OpenFile(target, flag, 0644)
	if err != nil {
		return WriteResult{Success: false, Error: err.Error()}
	}
	defer f.Close()

	n, err := f.WriteString(content)
	if err != nil {
		return WriteResult{Success: false, Error: err.Error()}
	}

	return WriteResult{
		Success:      true,
		Path:         filepath.ToSlash(relPath),
		BytesWritten: n,
		Mode:         modeStr,
	}
}

func (v *VaultManager) SearchVault(query string, maxResults int) []SearchResult {
	if maxResults <= 0 {
		maxResults = 15
	}
	queryLower := strings.ToLower(query)
	var results []SearchResult

	allowedExts := map[string]bool{
		".md":   true,
		".csv":  true,
		".txt":  true,
		".json": true,
	}

	filepath.Walk(v.RootDir, func(p string, info os.FileInfo, err error) error {
		if err != nil || len(results) >= maxResults {
			return nil
		}

		name := info.Name()
		if info.IsDir() {
			if strings.HasPrefix(name, ".") || name == "__pycache__" || name == "node_modules" {
				return filepath.SkipDir
			}
			return nil
		}

		ext := strings.ToLower(filepath.Ext(p))
		if !allowedExts[ext] {
			return nil
		}

		rel, err := filepath.Rel(v.RootDir, p)
		if err != nil {
			return nil
		}
		relSlash := filepath.ToSlash(rel)

		// Match filename
		if strings.Contains(strings.ToLower(name), queryLower) {
			results = append(results, SearchResult{
				Path:      relSlash,
				MatchType: "filename",
				Snippet:   fmt.Sprintf("Filename matches '%s'", query),
			})
			if len(results) >= maxResults {
				return filepath.SkipDir
			}
			return nil
		}

		// Match file content
		f, err := os.Open(p)
		if err != nil {
			return nil
		}
		defer f.Close()

		scanner := bufio.NewScanner(f)
		lineNum := 0
		for scanner.Scan() {
			lineNum++
			line := scanner.Text()
			if strings.Contains(strings.ToLower(line), queryLower) {
				snippet := strings.TrimSpace(line)
				if len(snippet) > 160 {
					snippet = snippet[:160] + "..."
				}
				results = append(results, SearchResult{
					Path:      relSlash,
					MatchType: "content",
					Line:      lineNum,
					Snippet:   snippet,
				})
				if len(results) >= maxResults {
					return filepath.SkipDir
				}
				break
			}
		}

		return nil
	})

	return results
}

File: app/api/vault.py
"""
Vault Manager for English Learning App
Provides safe local file access to Obsidian/local markdown/CSV vaults.
"""

import os
import glob
from pathlib import Path
from typing import List, Dict, Any, Optional

DEFAULT_VAULT_ROOT = os.path.abspath(
    os.path.join(os.path.dirname(__file__), "..", "..")
)

class VaultManager:
    def __init__(self, root_path: Optional[str] = None):
        self.root_path = Path(root_path or DEFAULT_VAULT_ROOT).resolve()

    def set_root(self, path: str) -> bool:
        resolved = Path(path).resolve()
        if resolved.exists() and resolved.is_dir():
            self.root_path = resolved
            return True
        return False

    def _resolve_safe_path(self, rel_path: str) -> Path:
        clean = rel_path.strip().lstrip("/\\")
        target = (self.root_path / clean).resolve()
        try:
            target.relative_to(self.root_path)
        except ValueError:
            raise PermissionError(f"Access denied: path '{rel_path}' is outside vault root.")
        return target

    def list_files(self, subpath: str = "", max_depth: int = 3) -> List[Dict[str, Any]]:
        target_dir = self._resolve_safe_path(subpath)
        if not target_dir.exists() or not target_dir.is_dir():
            return []

        results = []
        base_depth = len(target_dir.parts)

        for root, dirs, files in os.walk(target_dir):
            current_depth = len(Path(root).parts) - base_depth
            if current_depth >= max_depth:
                dirs.clear()
                continue

            # Skip hidden folders like .git, .agents, .gemini
            dirs[:] = [d for d in dirs if not d.startswith(".") and d != "__pycache__"]

            for f in sorted(files):
                if f.startswith("."):
                    continue
                full_path = Path(root) / f
                try:
                    rel_to_vault = str(full_path.relative_to(self.root_path))
                    stat = full_path.stat()
                    ext = full_path.suffix.lower()
                    results.append({
                        "name": f,
                        "path": rel_to_vault,
                        "size": stat.st_size,
                        "modified": int(stat.st_mtime),
                        "type": ext.lstrip(".") if ext else "file",
                        "is_markdown": ext == ".md",
                        "is_csv": ext == ".csv"
                    })
                except Exception:
                    continue

        return results

    def read_file(self, rel_path: str, max_bytes: int = 100000) -> Dict[str, Any]:
        target = self._resolve_safe_path(rel_path)
        if not target.exists():
            return {"error": f"File '{rel_path}' does not exist."}
        if not target.is_file():
            return {"error": f"'{rel_path}' is a directory, not a file."}

        try:
            content = target.read_text(encoding="utf-8", errors="replace")
            if len(content) > max_bytes:
                content = content[:max_bytes] + f"\n\n... [Content truncated, total {len(content)} characters] ..."
            return {
                "path": rel_path,
                "name": target.name,
                "content": content,
                "size": target.stat().st_size
            }
        except Exception as e:
            return {"error": f"Failed to read file: {str(e)}"}

    def write_file(self, rel_path: str, content: str, append: bool = False) -> Dict[str, Any]:
        target = self._resolve_safe_path(rel_path)
        try:
            target.parent.mkdir(parents=True, exist_ok=True)
            mode = "a" if append else "w"
            with open(target, mode, encoding="utf-8") as f:
                f.write(content)
            return {
                "success": True,
                "path": rel_path,
                "bytes_written": len(content.encode("utf-8")),
                "mode": "append" if append else "overwrite"
            }
        except Exception as e:
            return {"error": f"Failed to write file: {str(e)}"}

    def search_vault(self, query: str, max_results: int = 15) -> List[Dict[str, Any]]:
        query_lower = query.lower()
        results = []

        allowed_exts = {".md", ".csv", ".txt", ".json"}
        for root, dirs, files in os.walk(self.root_path):
            dirs[:] = [d for d in dirs if not d.startswith(".") and d != "__pycache__"]
            for f in files:
                ext = Path(f).suffix.lower()
                if ext not in allowed_exts:
                    continue
                file_path = Path(root) / f
                try:
                    rel_path = str(file_path.relative_to(self.root_path))
                    # Check filename match
                    if query_lower in f.lower():
                        results.append({
                            "path": rel_path,
                            "match_type": "filename",
                            "snippet": f"Filename matches '{query}'"
                        })
                        if len(results) >= max_results:
                            return results
                        continue

                    # Check text content match
                    with open(file_path, "r", encoding="utf-8", errors="ignore") as stream:
                        for line_idx, line in enumerate(stream):
                            if query_lower in line.lower():
                                results.append({
                                    "path": rel_path,
                                    "match_type": "content",
                                    "line": line_idx + 1,
                                    "snippet": line.strip()[:200]
                                })
                                if len(results) >= max_results:
                                    return results
                                break
                except Exception:
                    continue

        return results
